### PR TITLE
Exact scalar reduction to prevent underflows/overflows of int32

### PR DESCRIPTION
**Pull request by rafaelha**
This pull request refactors the handling of exact scalar arithmetic in the codebase, specifically focusing on the normalization and reduction of coefficients during sum and product operations. The main changes remove the explicit `.reduce()` method in favor of integrating reduction logic directly into arithmetic operations, leading to more numerically stable calculations.

**In particular, before this PR, some large diagrams would encounter underflow errors.**

Unfortunately, scalar reduction impacts performance. As a single data point: Execution time for the MSD d=3 circuit with 16 T gates goes from 2.4us to 4.4us on GH200.

Not every simulation requires high precision (this depends on the number of samples and the number of qubits in the autoregressive chain). In a future PR, we may consider adding a full fp32 option, which could give about one order of magnitude speedup.

### CODE REVIEW

File: src/tsim/core/exact_scalar.py [hunk 1]
```diff
@@ -77,45 +122,14 @@ def __mul__(self, other: "ExactScalarArray") -> "ExactScalarArray":
...
-        return ExactScalarArray(summed_coeffs, min_power.squeeze(-1))
+        """Sum elements along the last scalar axis using normalized pairwise adds."""
+        scanned_power, scanned_coeffs = lax.associative_scan(
+            _scalar_add_with_power, (self.power, self.coeffs), axis=-1
+        )
+        result_power = jnp.take(scanned_power, indices=-1, axis=-1)
+        result_coeffs = jnp.take(scanned_coeffs, indices=-1, axis=-2)
+        return ExactScalarArray(result_coeffs, result_power)
 
     def prod(self, axis: int = -1) -> "ExactScalarArray":
```
[the remark this answers is not in the document]
Comment (rafaelha) [reply src/tsim/core/exact_scalar.py]: Changed default to -2
